From 5dedc074b08c5676ff2d41b0b1a3202b2dcd7f57 Mon Sep 17 00:00:00 2001
From: Xiongfeng Wang <wangxiongfeng2@huawei.com>
Date: Fri, 3 Dec 2021 18:15:56 +0800
Subject: [PATCH] arm64: kexec: only clear EOI for SDEI in NMI context
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 5dedc074b08c5676ff2d41b0b1a3202b2dcd7f57


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I4KCU2
CVE: NA

----------------------------------------

We need to clear EOI for the secure timer only when we panic from
sdei_handler. If we clear EOI for the secure timer in normal panic
routiue, it has no bad effect on Hi1620, but it may cause undefine
behavior on Hi1616. So add a check for NMI context before we clear EOI
for the secure timer.

Fixes: dd397d5febc4("sdei_watchdog: clear EOI of the secure timer before
kdump")

Signed-off-by: Xiongfeng Wang <wangxiongfeng2@huawei.com>
Reviewed-by: Wei Li <liwei391@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Xiongfeng Wang <wangxiongfeng2@huawei.com>
Reviewed-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm64/kernel/machine_kexec.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/kernel/machine_kexec.c b/arch/arm64/kernel/machine_kexec.c
index 53def49c2ea3..0b09ee49cfba 100644
--- a/arch/arm64/kernel/machine_kexec.c
+++ b/arch/arm64/kernel/machine_kexec.c
@@ -262,7 +262,8 @@ void machine_crash_shutdown(struct pt_regs *regs)
 	 * interrupt failed to trigger in the second kernel. So we clear eoi
 	 * of the secure timer before booting the second kernel.
 	 */
-	sdei_watchdog_clear_eoi();
+	if (in_nmi())
+		sdei_watchdog_clear_eoi();
 
 	/* for crashing cpu */
 	crash_save_cpu(regs, smp_processor_id());
-- 
2.26.2

